What is a codec server (#1241)
temporalio/documentation: docs/concepts/what-is-a-codec-server.md

---
id: what-is-a-codec-server
title: What is a Codec Server?
sidebar_label: Codec Server
description: Explanation and implementation of a remote encryption/decryption server.
---

> A Codec Server is an optional feature for altering Payloads returned from a default Data Converter.

Reviewer: ```suggestion
A Codec Server is an optional feature that carries out additional levels of encoding/decoding to Payloads that are handled by tctl or the Web UI.
The Web UI and tctl both use a default Data Converter, which is only capable of serialization.
```

- Link to Payload (I think we should have a "What is a Payload" term page, and define it - because this keeps coming up )
- Link to Data Converter
- Link to tctl
- Link to Web UI

---
id: what-is-a-codec-server
title: What is a Codec Server?
sidebar_label: Codec Server
description: Explanation and implementation of a remote encryption/decryption server.
---

A Codec Server is an optional feature that carries out additional levels of encoding/decoding to Payloads that are handled by [tctl](/docs/tctl/index) or the [Web UI](/docs/web-ui/how-to-use-a-list-filter-in-the-temporal-web-ui).

The Web UI and tctl both use a default [Data Converter](/docs/concepts/what-is-a-data-converter), which is only capable of serialization.

> Codec Servers can be used to encrypt, compress, and change the format of a Payload object. These measures further secure your data while formatting it to your preference.

Reviewer: excellent.

> ## Use Case - tctl

Reviewer: ```suggestion
## Use case - tctl
```

Suppose that a developer wants to view Workflow History. They can accomplish this with the `workflow show` command with tctl. A Payload is retrieved and sent to a default Data Converter, which converts the Payload object into a human-readable format. Suppose it wasn't what you wanted.

> Codec Servers allow further customization for Payload objects. The default Data Converter sends the Payload to a given endpoint, and receives a decoded Payload if the API returns a successful result. The Data Converter passes this back to the Client for viewing.

Reviewer: Try to adjust the tone here - and be more direct.
For example, I image this would start out more like:

> You can use the [tctl workflow show](link to command) command to view a Workflow Execution Event History. When this happens, the Payload is deserialized by tctl's default Data Converter...

---
id: what-is-a-codec-server
title: What is a Codec Server?
sidebar_label: Codec Server
description: Explanation and implementation of a remote encryption/decryption server.
---

A Codec Server is an optional feature used to encrypt and decrypt Payloads for the WebUI.

![](/img/remote-codec-server-problem.svg)

> Each Client has a built-in [Data Converter](/docs/concepts/what-is-a-data-converter). However, the pre-built tctl and WebUI binaries won't be able to decrypt the serialized Payloads.

Reviewer: ```suggestion
Each Temporal Client has a built-in [Data Converter](/docs/concepts/what-is-a-data-converter) that serializes Payloads.
However, the pre-built tctl and WebUI binaries are unable to decrypt the serialized Payloads.
```

We are using the terms "serialize", "encode", and "encrypt" -
Let's make sure we are using them correctly.

Serializing is about structuring data
Encoding is a general term for transforming the data
Encrypting is about transforming the data using a key and algorithm

The setting in the Web UI is called "Data Encoder".
And then you can use either the tctl Plugin or the Codec Server.

Obviously the WebUI can handle the serialized data?

So, it seems like the Codec Server is specifically for any handling any other level of encoding/decoding that is happening to the data - including encryption?

![](/img/remote-codec-server-solution.svg)

With a Codec Server, the Client's Data Converter can send an HTTP API request to the codec to decrypt a Payload.
This decrypted data is passed back to the WebUI.

Decrypted data can also be passed back through the Codec Server for encryption. Then, when passed through the Client's Data Converter, it can be deserialized before being sent back to the Temporal Cluster.

![](/img/remote-codec-server-diagram.svg)

> ## Configuration

Reviewer: I don't think we need this header.

What we need is an additional "how to develop a Codec Server" or "How to run a Codec Server" page -

> Before running a Codec Server, make sure that a [Temporal service](https://docs.temporal.io/application-development-guide/#run-a-dev-cluster) is running in your SDK of choice.

Reviewer: ```suggestion
Before running a Codec Server, make sure that a [Temporal Cluster](https://docs.temporal.io/application-development-guide/#run-a-dev-cluster) is running.
```

Clusters are not run using an SDK. Worker Processes are run using an SDK.

But I am curious why this is important?

Its either tctl or the Web UI that calls the Code Server - so why is this an important call out?

If you are using Go, feel free to base your remote Codec Server on the [temporalio/samples-go](https://github.com/temporalio/samples-go) repository.

> For other languages, please refer to the diagrams in the Go sample to construct the Codec Server.

Reviewer: This is incredibly unspecific.

Per my comment above - I think we need to show exactly how to set this up as part of a how-to.

---
id: what-is-a-codec-server
title: What is a Codec Server?
sidebar_label: Codec Server
description: Explanation and implementation of a remote encryption/decryption server.
---

> A Codec Server is an optional implementation for your [Temporal Platform](/docs/concepts/what-is-the-temporal-platform). With it, deserialized Payloads can have their data displayed on your WebUI.

Reviewer 2: I'm inclined to change "Payloads" to "Payload objects", at least at this first instance, unless we'll have "Payloads" as a first-class Temporal term. (Or lowercase "Payload[s]".) However, I'm not the expert. @flossypurse, any opinion?

Reviewer: On the other hand.
Might be a worthwhile term for us.

## Purpose

![](/img/remote-codec-server-problem.svg)

By default, the Temporal Platform provides Clients to receive information from the tctl, WebUI, and running Workers. Each Client has a built-in [Data Converter](/docs/concepts/what-is-a-data-converter).

Data Converters serialize and deserialize between a language's native data types and raw Payloads received from the Temporal Cluster. However, the pre-built tctl and WebUI binaries won't be able to decrypt the Payloads.

![](/img/remote-codec-server-solution.svg)

> With a Codec Server, the Client's Data Converter can send an HTTP API to request the codec to decrypt a given Payload. This decrypted data is passed back to a user-facing interface (whether through the command line or by WebUI) for viewing by the developer.

Reviewer 2: ```suggestion
With a Codec Server, the Client's Data Converter can send an HTTP API request to the codec to decrypt a Payload.
This decrypted data is passed back to a user-facing interface (through the command line or by WebUI) for viewing by the developer.
```

Decrypted data can also be passed back through the Codec Server for encryption. Then, when passed through the Client's Data Converter, it can be deserialized before being sent back to the Temporal Cluster.

> The Codec Server can only receive HTTP APIs from Clients. The WebUI and tctl cannot send or receive anything directly to the Codec Server.

Reviewer 2: ```suggestion
The Codec Server can receive HTTP API calls only from Clients.
The WebUI and tctl cannot send anything directly to or receive anything directly from the Codec Server.
```
If you can think of a cleaner way of avoiding "receive to", be my guest!

## Configuration

Codec Servers are initiated in a similar manner to a Temporal Worker.

Before running a Codec Server, make sure that a [Temporal service](https://docs.temporal.io/application-development-guide/#run-a-dev-cluster) is running in your SDK of choice.

> If you are using Go, feel free to base your remote Codec Server off of [this existing repo](https://github.com/temporalio/samples-go). For other languages, please refer to the diagrams in the Go sample to construct the Codec Server.

Reviewer 2: ```suggestion
If you are using Go, feel free to base your remote Codec Server on the [temporalio/samples-go](https://github.com/temporalio/samples-go) repository.
For other languages, please refer to the diagrams in the Go sample to construct the Codec Server.
```
Link text change per https://developers.google.com/style/link-text#write-link-text.

---
id: what-is-a-codec-server
title: What is a Codec Server?
sidebar_label: Codec Server
description: Explanation and implementation of a remote encryption/decryption server.
---

A codec server provides a means of securely decrypting payload data for display by Temporal user interfaces.

> Data from the customer’s remote server is decrypted via a codec server before it can be viewed on the WebUI. Conversely, data from the WebUI is encrypted through the codec server before being passed along to the database. Therefore, it is more secure for Temporal to store encrypted data.

Reviewer: I don't understand what a "remote server" is in this context.

I also am unclear on what the "database" is in this context as well... 

In the diagrams I see below - neither mention a remote server or a database.

The web UI only really communicates with the Temporal Cluster (lets change to that vs Temporal Server)

The Codec Server is there to encode/decode data coming to/from the Cluster - at least, that is my understanding.

Any chance we can improve?

Codec servers can handle authentication and authorization in a similar manner used by Temporal Clusters. In the event of a data breach, any information uncovered would be indecipherable to bad actors.

## Purpose

> In application code DataConverters can be configured to use a PayloadCodec to encrypt data before it’s sent to Temporal.

Author: That's how they're referred to in other documents

> In application code DataConverters can be configured to use a PayloadCodec to encrypt data before it’s sent to Temporal.

Author: I see now. Will revise

![](/static/img/remote-codec-server-problem)

However, the pre-built tctl and WebUI binaries use a default DataConverter that excludes your PayloadCodec.

Because of this, the tctl and the WebUI will not be able to show decrypted data—at least, not without help.

![](/static/img/remote-codec-server-solution)

> This is where a remote Codec Server comes into play. The server exposes your PayloadCodec’s encode/decode methods via HTTP interface. This allows tctl and the WebUI to use the PayloadCodec to encrypt and decrypt data as needed.

Reviewer 3: ```suggestion
This is where a Codec Server comes into play. The server exposes your PayloadCodec’s encode/decode methods via HTTP interface. This allows tctl and the WebUI to use the PayloadCodec to encrypt and decrypt data as needed.
```

---
id: what-is-a-codec-server
title: What is a Codec Server?
sidebar_label: Codec Server
description: Explanation and implementation of a remote encryption/decryption server.
---

> A Codec Server is an optional implementation for your [Temporal Platform](/docs/concepts/what-is-the-temporal-platform). With it, deserialized Payloads can have their data displayed on your WebUI.

Reviewer: Maybe something like...

"A Codec Server is an optional feature used to encrypt and decrypt data payloads for the Web UI."

> ## Purpose

Reviewer: Not a fan of this header.
I don't think we need one yet.

![](/img/remote-codec-server-problem.svg)

> By default, the Temporal Platform provides Clients to receive information from the tctl, WebUI, and running Workers. Each Client has a built-in [Data Converter](/docs/concepts/what-is-a-data-converter).

Reviewer: The first sentence in this paragraph doesn't make any sense to me.

> On the Temporal Platform, you can use the WebUI or the [tctl plugin](/docs/tctl/index) to request data from the Temporal Cluster. This request is passed through a client's [Data Converter](/docs/concepts/what-is-a-data-converter), which sends and receives API calls from the Remote Codec Server. These calls are then passed from the client to the cluster, where a successful request passes an encoded payload back through the platform.

Reviewer: A Data Converter resides in the Codec Server.

The Web UI makes a call to the Codec Server only when it needs to encode/decode data  before/after API calls to the Cluster.
@robholland ^ is this accurate?

Author: Thank you, Rob! This really clears things up

Author: @robholland Should I correct the diagrams?

> Data Converters serialize and deserialize between a language's native data types and raw Payloads received from the Temporal Cluster. However, the pre-built tctl and WebUI binaries won't be able to decrypt the Payloads.

Reviewer: I think you can clean this up and make it part of the first several sentences in the page.

---
id: what-is-a-codec-server
title: What is a Codec Server?
sidebar_label: Codec Server
description: Explanation and implementation of a remote encryption/decryption server.
---

A Codec Server is a feature that can perform additional levels of encoding and decoding on Payloads that are handled by [tctl](/docs/tctl/index) or the [Web UI](/docs/web-ui/how-to-use-a-list-filter-in-the-temporal-web-ui).

The Web UI and tctl both use a default [Data Converter](/docs/concepts/what-is-a-data-converter), which is capable of serialization only.

Codec Servers can encrypt, compress, and change the format of a Payload object.
These measures can further secure your data.

> ![](/img/tctl-diagram-codec-server.svg)

Reviewer: This is a pretty good generic diagram...

I think there could be more labels that clear up the "Encryption" item and relationship with keys.

But it passes muster for the moment. Nice work.

---
id: what-is-a-codec-server
title: What is a Codec Server?
sidebar_label: Codec Server
description: Explanation and implementation of a remote encryption/decryption server.
---

A Codec Server is a feature that can perform additional levels of encoding and decoding on Payloads that are handled by [tctl](/tctl/index) or the [Web UI](/web-ui/how-to-use-a-list-filter-in-the-temporal-web-ui).

- [How to set up a Codec Server](/clusters/how-to-set-up-codec-server/)

The Web UI and tctl both use a default [Data Converter](/concepts/what-is-a-data-converter), which is capable of serialization only.

Codec Servers can encrypt, compress, and change the format of a Payload object.
These measures can further secure your data.

![](/img/tctl-diagram-codec-server.svg)

## Use case: tctl

Suppose that you want to view Workflow History.
This information needs to be decoded before it can be viewed.

You can use [tctl workflow showid](/tctl/workflow/showid) to view a Workflow Execution Event History.

```bash
    tctl workflow showid <workflowID>
```

> With a Codec Server, the Payload is decoded after being deserialized by tctl's default Data Converter. The default Data Converter sends the Payload to a given endpoint, and receives a decoded Payload if the API returns a successful result.

Reviewer 3: The Payload is decoded before being deserialised. For example if the Payload has been encrypted, it would need to be decrypted before the DataConverter can deserialise it.
